client/src/util/centroid.js

```diff
@@ -1,61 +1,125 @@
 import quantile from "./quantile";
+import { memoize } from "./dataframe/util";
+import { unassignedCategoryLabel } from "../globals";
 
 /*
   Centroid coordinate calculation
 */
 
-/* Unused - please cleanup
-const calcMeanCentroid = (world, annoName, annoValue, layoutDimNames) => {
-  const centroid = { x: 0, y: 0, size: 0 };
-  const annoArray = world.obsAnnotations.col(annoName).asArray();
-  const layoutXArray = world.obsLayout.col(layoutDimNames[0]).asArray();
-  const layoutYArray = world.obsLayout.col(layoutDimNames[1]).asArray();
-
-  for (let i = 0, len = annoArray.length; i < len; i += 1) {
-    if (annoArray[i] === annoValue) {
-      centroid.x += layoutXArray[i];
-      centroid.y += layoutYArray[i];
-      centroid.size += 1;
-    }
-  }
-
-  if (centroid[2] !== 0) {
-    centroid.x /= centroid.size;
-    centroid.y /= centroid.size;
-  }
+/* 
+  calcMedianCentroid goes through a given metadata category
+  fetches each cell's coordinates grouping by category value.
 
-  return [centroid.x, centroid.y];
-};
+  It then calculates the median value and puts that in the array
 */
 
-const calcMedianCentroid = (world, annoName, annoValue, layoutDimNames) => {
-  const centroidX = [];
-  const centroidY = [];
-  let hasFinite = false;
+const calcMedianCentroid = (
+  obsAnnotations,
+  obsLayout,
+  categoryName,
+  layoutDimNames,
+  categoricalSelection,
+  schemaObsByName
+) => {
+  const categoryArray = obsAnnotations.col(categoryName).asArray();
+
+  const layoutXArray = obsLayout.col(layoutDimNames[0]).asArray();
+  const layoutYArray = obsLayout.col(layoutDimNames[1]).asArray();
+  const coordinates = new Map();
+
+  // Iterate over all the cells in the category
+  for (let i = 0, len = categoryArray.length; i < len; i += 1) {
+    const categoryValue = categoryArray[i];
+
+    // Get the index of the categoryValue within the category
+    // If the category is truncated and this value is removed,
+    // it will not be assigned a category value and will not be
+    // labeled on the graph
+    const categoryValueIndex = categoricalSelection[
```

Comment:
This could be optimized significantly if you pull all of this setup code, which does not change as you iterate through the loop, out into a setup pass.  Use the schema - it has all of the information already in it.

I think the following lines could be done once, rather than for each row/cell:  lines 38-63

Reply:
Per our offline conversation, this optimization can be done as a later PR (ie, land this thing!)    Please make sure you capture in an issue, linked from this PR, so we don't forget about it.